Engine/Platform/System/Win32/CYBWin32File.cpp

```diff
@@ -0,0 +1,170 @@
+{
+	if (AMode == Mode::READ && AMethod == Method::TRUNCATE)
+		throw Exception::Violation(Exception::Violation::INVALID_PARAMETERS);
+
+	auto& K32(Core().FModuleManager.FK32);
+
+	const auto DesiredAccess([&]() -> DWORD {
+		switch (AMode)
+		{
+		case Mode::READ:
+			return GENERIC_READ;
+		case Mode::WRITE:
+			return GENERIC_WRITE;
+		case Mode::READ_WRITE:
+			return GENERIC_READ | GENERIC_WRITE;
+		default:
+			throw Exception::Violation(Exception::Violation::INVALID_ENUM);
+		}
+	}()), CreationDisposition([&]() -> DWORD {
+		switch (AMethod) {
+		case Method::ANY:
+			return OPEN_ALWAYS;
+		case Method::CREATE:
+			return CREATE_NEW;
+		case Method::EXIST:
+			return OPEN_EXISTING;
+		case Method::TRUNCATE:
+			return CREATE_ALWAYS;
+		default:
+			throw Exception::Violation(Exception::Violation::INVALID_ENUM);
+		}
+	}());
+
+	const auto FilePath(FPath.WidePath().WString());
+																								//Maintain POSIX compatibility, no locking
+	FHandle = K32.Call<Modules::Kernel32::CreateFileW>(FilePath, DesiredAccess, static_cast<DWORD>(FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE), nullptr, CreationDisposition, static_cast<DWORD>(FILE_ATTRIBUTE_NORMAL), nullptr);
+
+	if (FHandle == INVALID_HANDLE_VALUE) {
+		const auto Error(K32.Call<Modules::Kernel32::GetLastError>());
+		API::Assert::NotEqual<DWORD>(Error, ERROR_INVALID_PARAMETER);
+
+		//I know this could be a switch but CodeCoverage.exe hates it
+
+		if (Error == ERROR_FILE_NOT_FOUND || Error == ERROR_PATH_NOT_FOUND || Error == ERROR_INVALID_NAME)
+			throw Exception::SystemData(Exception::SystemData::FILE_NOT_FOUND);
+		if (Error == ERROR_FILE_EXISTS)
+			throw Exception::SystemData(Exception::SystemData::FILE_EXISTS);
+		if (Error == ERROR_ACCESS_DENIED) {
+			//Translate properly to the directory error
+			//We don't care how screwed up FPath is, it has a different exception spec to us
+			//DO NOT LET IT PROPAGATE
+			bool IsDir;
+			try {
+				IsDir = FPath.IsDirectory();
+			}
+			catch (CYB::Exception::Base& AException) {
+				API::Assert::Equal(AException.FLevel, Exception::Base::Level::SYSTEM_DATA);
+				IsDir = false;
+			}
+			if (IsDir)
+				throw Exception::SystemData(Exception::SystemData::FILE_EXISTS);
+		}
+		if (AMode == Mode::READ)
+			throw Exception::SystemData(Exception::SystemData::FILE_NOT_READABLE);
+		throw Exception::SystemData(Exception::SystemData::FILE_NOT_WRITABLE);
+
+	}
+	
+	if (AMethod == Method::ANY) {
+		if (K32.Call<Modules::Kernel32::GetLastError>() == ERROR_ALREADY_EXISTS)
+			FOpenMethod = Method::EXIST;
+		else
+			FOpenMethod = Method::CREATE;
+	}
+	else
+		FOpenMethod = AMethod;
+}
+
+CYB::Platform::System::Implementation::File::File(File&& AMove) noexcept :
+	FHandle(AMove.FHandle)
+{
+	AMove.FHandle = INVALID_HANDLE_VALUE;
+}
+
+void CYB::Platform::System::Implementation::File::Close(void) const noexcept {
+	if (FHandle != INVALID_HANDLE_VALUE)
+		Core().FModuleManager.FK32.Call<Modules::Kernel32::CloseHandle>(FHandle);
+}
```

Comment:
Now that I see it, I realize the declaration of close and destructor can be moved into the generic level